Hook wrapping without `_Result`

bluetech opened the pull request:
This implements the idea from #260. I think it makes pluggy more elegant, and also provides a backward compatible fix for #244 (replaces #257).

bluetech commented:
In #260, @maxnikulin argued against using a "plain" generator function for this, preferring a new `@hookimpl(wrapper=True)` instead, to make it explicit.

My opinion is that implicit is fine, because that's how python functions themselves work, and how e.g. `pytest.fixture` works, and I think they're mostly fine. But I'm willing to be outvoted on this.

RonnyPfannschmidt reviewed (approved):
This looks nicer than what we have
It Will make async a little Harder, but reasonably so

bluetech commented:
>  It Will make async a little Harder, but reasonably so

Seems like async generators have the necessary `asend`/`athrow`, the only thing they're missing is `return` support. I wonder why they don't support that.

bluetech commented:
Thanks for the reviews @RonnyPfannschmidt.

I plan to:
- Extract to `force_exception` change to its own PR with own changelog.
- Wait a week or two to allow for discussion.
- Try to port pytest to detect any issues and measure performance in a realistic setting.

bluetech commented:
I've made the following performance measurements:

1. Pluggy benchmark - old pluggy/new pluggy - neutral
2. Pytest benchmarks - old pytetst+old pluggy/new pytest+new pluggy - neutral
3. Pytest benchmarks - old pytetst+old pluggy/old pytest+new pluggy - neutral

I also made an experiment where I removed `hookwrapper=True` support entirely from `_multicall`  (removing `_Result` etc.) and this gets a ~15% improvement in pluggy benchmark. But we don't have to drop old-style wrappers to get this improvement; it is easy to add a fast path which specializes for only new-style case - see [here](https://github.com/bluetech/pluggy/commits/un-result-fast-path) (better to view ignoring whitespace). Will maybe submit this as a separate PR if this one gets merged.

RonnyPfannschmidt commented:
How expensive would it be to have legacy hook wrappers be implemented in terms off The new wrappers thus making them slower but allowing the new wrappers also Bugfix the exception escape

bluetech commented:
I'm not sure, but I generally think it'd be better to keep the existing hookwrapper=True semantics as-is, to avoid breaking things (https://xkcd.com/1172/ situation). Switching to the new style would also be opting in to the new semantics.

nicoddemus commented:
Great work @bluetech! 

The idea seems great, I will take a look in the code in the next few days! :+1:

nicoddemus reviewed (approved):
Looks great, thanks @bluetech!

bluetech commented:
Thanks for the reviews!

After taking another look at #257, I've made a few changes:

- Added the `teardown.close()` bit -- it's good to have. Added it only for the new style so as not to touch old style.
- Ported the tests from there.
- Added @maxnikulin as co-author

bluetech commented on docs/index.rst:
```diff
@@ -362,13 +362,73 @@ For another example see the :ref:`pytest:plugin-hookorder` section of the
+        """Wrap calls to ``setup_project()`` implementations which
+        should return json encoded config options.
+        """
+        # get initial default config
```
Oops I forgot to handle this -- fixing now.
